### ISSUE TITLE
weird semantics for `exit` on multi-domain programs

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#11176]
(cc @kayceesrk @Engil)

In reading the code of `Domain.at_exit` to check that the order of execution is as described in #11175, I noticed that `Domain.at_exit f` calls `Stdlib.at_exit f` -- as also documented. This feels wrong to me:
- it means that in general a `Domain.at_exit`-passed cleanup function will be called several times, first when the domain exits and then at program shutdown, which sounds unexpected and probably a bug
- it means that we cannot use `Domain.at_exit` to cleanup a domain-local resource, or that the function itself has to guard against being called from another domain.

I would hazard the guess that this choice was is motivated by the fact that the "usual" Domain.at_exit handler is not called on the first domain (I thought it was related to the fact that `terminate_domain` is not called, but in fact `Domain.at_exit` is handled purely on the OCaml side by `Domain.spawn`, which is not used for the first domain.), and we only tested `Domain.at_exit` on the first domain or with idempotent functions.

Here would be my expectation for `Domain.at_exit`:
- the passed callbacks are called at most once
- they always get called exactly once when the program terminates correctly
- they are always executed on the domain that registered them with `Domain.at_exit`
- same guarnatees for `Stdlib.at_exit`, but it's not specified which domain will run the callback. (Not necessarily the registering domain.)

**Comment by gasche** [comment]
(Also interested, I guess: @favonia, @gadmm.)

**Comment by gasche** [comment]
Currently, if I read the code correctly, if a domain calls `Stdlib.exit`, then that domain will run the `Domain.at_exit` handlers of all domains (through the fact that they were also registed with `Stdlib.at_exit`), while those domains are executing concurrently (including potentially terminating and calling those handlers again), and at some later point the exiting domain will abruptly terminate all other domains.

**Comment by abbysmal** [comment]
Hello,

Indeed there's an odd discrepancy here.

> the passed callbacks are called at most once

I think it would make sense for `Domain.at_exit` to be a Domain local callback, and that should enforce this property?

> they always get called exactly once when the program terminates correctly
> they are always executed on the domain that registered them with Domain.at_exit

Likewise with the above suggestion: I think a Domain local `Domain.at_exit` would enforce these invariants.

> same guarnatees for Stdlib.at_exit, but it's not specified which domain will run the callback. (Not necessarily the registering domain.)

I think `Stdlib.at_exit` is already protected from the "re-running" a callback issue in the wrapping code (using the `f_yet_to_run` atomic variable).

I have a feeling things may be ok if:

- `Domain.at_exit` are domain-local callbacks, with the same "no-re-run" protection as `Stdlib.exit`.
- `Domain.at_exit` and `Stdlib.at_exit` callbacks are different. (ie. we do not register anything implicitly to the `Stdlib.at_exit` set.)
- Domain 0 is also outfitted with a check for `Domain.at_exit` during termination.
- Agreed that we can hardly force `Stdlib.at_exit` callbacks to be executed by a specific domain only.

One thing I am not sure how to solve with this design, is how to make sure that `Domain.at_exit` callbacks are being run after calling `Stdlib.exit`. With my suggested design, the domain-local callbacks would not be run (as they are domain local) by the domain calling `Stdlib.exit`.

Maybe it is fine to consider that calling `Stdlib.exit` with multiple domains could implies improper resources terminations (ie. not running the domain local callbacks), or maybe this design is just too weak.

I have a feeling that conflating `Stdlib.exit` (the program exiting, killing all threads in the process) and Domain termination is the real source of headscratchers here, but I cannot think yet of a nice enough design covering all corner cases.

**Comment by gadmm** [comment]
Yes, the programmer is responsible for only calling `Stdlib.exit` after other domains have terminated if they want proper cleanup. The rest ventures on the question of thread cancellation. There is nothing special to do in Stdlib.exit apart from making sure that the domain-local cleanup of the last domain can be done in sensible order, and making sure that the programmer is aware of the difficulty. This means that the documentation should look like the Rust equivalent: https://doc.rust-lang.org/std/process/fn.exit.html.

**Comment by xavierleroy** [comment]
> the programmer is responsible for only calling Stdlib.exit after other domains have terminated if they want proper cleanup

It's worse than that: under Windows, when a thread terminates the whole program while other threads are still running, a crash (of the General Protection Fault kind) is often reported.   So, please, join your threads and your domains before exiting.  I spent much time adding join operations to the tests in our test suite just for this reason.  Also, I wish we had an API for domain / thread creation that forces the parent domain / thread to join before terminating, but it's probably too late for that.

I think we should shoot for the following:
- The domain that calls `Stdlib.exit` runs its `Domain.at_exit` finalizers for itself, then the `Stdlib.at_exit` finalizers.
- Actually, it's `Stdlib.do_at_exit` that should run the `Domain.at_exit finalizers`, then the `Stdlib.at_exit` finalizers.  This way, if the initial domain terminates by falling through, it runs `do_at_exit()` via `stdlib/std_exit.ml`, and domain finalizers will be run too.
- This opens the possibility that `Domain.at_exit` finalizers are executed multiple times, if a finalizer calls `Stdlib.exit`.  If we really care, we should make `Domain.at_exit` finalizers "run once", with the same trick used for `Stdlib.at_exit` finalizers.
- With the plan above, I don't think a special case for domain 0 is needed.

**Comment by gadmm** [comment]
Then `Stdlib.exit` raises an exception if not running on the last domain.

**Comment by gasche** [comment]
After sleeping on it, I realize that aspects of my original description are wrong: `Domain.at_exit f` is not meant to run a callback on only the calling domain, but to run a callback on *all* domains as they exit. In particular, protecting the callback against running several times does not fully make sense, as we do expect it to run on each domain. What we expect is that it will run only once per domain.

**Comment by gasche** [comment]
I can observe a strange behavior on the following test program, which calls `exit` from a non-0 domain (that is never joined):

```ocaml
let dom : unit Domain.t = Domain.spawn (fun () -> exit 0)
```

With the default settings this program always succeeds, but when run in cleanup mode (`OCAMLRUNPARAM=c`), this program will sometimes (once every 10 runs or so) fail with this error:

> Fatal error: a call to caml_shutdown has no corresponding call to caml_startup

If I add a call to `Domain.join`, this error seems to go away.

@gadmm and @xavierleroy have made several suggestions that would let us consider that this test program is undesirable/faulty, but it looks like it currently puts the runtime in a weird/inconsistent state, it would be better to have a clean error message instead.

If I understand correctly:
- @gadmm proposes to consider it a programming error if `exit` is called while *other* domains are still running
- concretely he suggests to have an explicit error calling `exit` from a non-zero domain (but didn't suggest checking whether all domains are still running)
- @xavierleroy has a proposal that would give a reasonable semantics to calling `exit` from a non-0 domain

In any case, no one seems to suggest that we should try to properly terminate other domains when `exit` happens. (We could have a stop-the-world section for this.) The Rust behavior explicitly points out that destructors are *not* called on `exit`, it's the user responsibility to only call it from the outermost layer of the program, and this is what @gadmm suggests here as well.

**Comment by gasche** [comment]
To be clear, the two points I initially raised were:

> - it means that in general a Domain.at_exit-passed cleanup function will be called several times, first when the domain exits and then at program shutdown, which sounds unexpected and probably a bug
> - it means that we cannot use Domain.at_exit to cleanup a domain-local resource, or that the function itself has to guard against being called from another domain.

Both remarks are invalid if we consider that `Domain.at_exit` will run its callback on all future domains, not just the current domain, so they are just a result of my misunderstanding.

I'm inclined to close this PR and maybe submit a few tests, but the discussion surfaced the question of what we want to do when `exit` is called from a non-0 domain, where apparently some margin for improvement remains.

**Comment by xavierleroy** [comment]
I'm increasingly confused by your description of the issues.

> the question of what we want to do when exit is called from a non-0 domain, where apparently some margin for improvement remains.

There are problems when `Stdlib.exit` is called from a domain while other domains are still running.  Whether `Stdlib.exit` is called from domain 0 or from another domain is irrelevant.

Actually it's slightly relevant: an uncaught exception in domain 0 causes `Stdlib.exit` to be called.  That's why I think it doesn't help to make `Stdlib.exit` raise an exception if other domains are still running:
```
uncaught exception -> Stdlib.exit -> uncaught exception "you can't exit" -> exception ignored or Stdlib.exit is entered again -> ...
```

Finally, the cleanup-at-exit mode `OCAMLRUNPARAM=c` was broken by the switch to Multicore OCaml, and probably still doesn't quite work, so I would not draw any conclusion from experiments ran in cleanup-at-exit mode.

**Comment by gasche** [comment]
Okay, so let's close this for now. I feel that things are not right, but it's not helpful unless one has clear specifications to propose and problem descriptions.

**Comment by gadmm** [comment]
You just found more issues. (As earlier, I react according to what I read here, as I did not spend time inspecting the code.)

* `Domain.at_exit` callbacks running on all domains is an issue. It does not make sense to me to run a clean-up function without knowing that a matching initialization has been performed. With `Domain.at_startup` running on all future domains, you can still call `at_exit` inside `at_startup` (and it would still make more sense than currently). With more context I now realize that this is what @Engil was proposing.

* According to @xavierleroy's description, there is a bug in the handling of uncaught exceptions on domain zero whereby `exit(2)` and `Stdib.at_exit` handlers can be called while other domains are still running. Maybe it should join the other domains after printing the unhandled exception message and before running `at_exit` handlers (this means switching their order). (Note that it does not exactly call `Stdlib.exit` after an uncaught exception.) Thus when calling `Stdlib.exit`, you would get an exception, which, if uncaught gives an error message before perhaps deadlocking.

Unlike a PR, an issue need not have a clear solution to be helpful. I think this should stay open.

**Comment by xavierleroy** [comment]
I think there are two issues here that are mostly independent but need to be discussed, preferably separately:

1. The semantics of `Domain.at_exit`.  Like @gadmm I find it very strange that an exit handler registered with `Domain.at_exit` applies to domains that are already running.  And I'm not sure it should apply to domains created in the future, instead of just to the calling domain.  More generally, the whole business of `Domain.at_first_spawn` / `Domain.at_startup` / `Domain.at_exit` is unclear to me.  Right now we have very limited uses within the standard library and the threads library.  I wonder if we should really expose these functions to users in 5.0, or keep them "for internal use only" until we understand better.

2. Whether `Stdlib.exit` should wait for all domains (and threads?) to terminate.  On the one hand it would give a nicer story w.r.t. resource management.  On the other hand, we don't have a domain/thread cancellation mechanism, and `Stdlib.exit` needs to be usable as an emergency stop mechanism.
3.

**Comment by gadmm** [comment]
Domain.at_exit on a per-domain basis is meaningful for the clean-up of domain-local resources. It is possible that during the migration to multicore, some calls to Stdlib.at_exit will be converted to Domain.at_exit.

Has anyone arguments for something else than having Domain.as_exit on a per-domain basis?

**Comment by sadiqj** [comment]
> I wonder if we should really expose these functions to users in 5.0, or keep them "for internal use only" until we understand better.

They do feel a little odd at the moment, even ignoring the "quirks" in their behaviour.

I think the motivating use-case is where you are managing some resource for all domains. The problem is I can't think of a way of doing that in a user library (not the stdlib) with the current API that doesn't race another library spawning a second domain.

> Has anyone arguments for something else than having Domain.as_exit on a per-domain basis?

I think for a _user_ API it being global is pretty surprising. For an internal API it makes more sense. For some context, we originally wanted `Domain.at_exit` for domain local storage but then came up with a better design.

If we continued to expose it and made it per-domain then expressing the use in https://github.com/ocaml/ocaml/blob/trunk/stdlib/format.ml#L1475 would become a little bit more convoluted.

**Comment by gadmm** [comment]
In this example how about something like `let () = Domain.at_exit f ; Domain.at_startup (fun () -> Domain.at_exit f)`? The line below it already assumes that no other domain has been spawned yet.

**Comment by kayceesrk** [comment]
I should say that `Domain.at_exit` was added specifically to address its only use in `stdlib/format.ml` and is not very well thought out beyond that one use case. I agree that the semantics of `Domain.at_exit` is odd at the moment and not the right one.  

I am not against hiding `Domain.at_first_spawn`, `Domain.at_startup` and `Domain.at_exit` to be internal-only in 5.0. I'd rather not expose this functionality than expose a broken one. It would be best if we can use it in `Format` module for now, and not expose it in the domain interface. Is there an easy way to do this? Alternatively, is there a different solution that doesn't use these functions for the problem in `Format` with multiple domains?

If we are to expose `Domain.at_exit`, then I agree that it should definitely be domain-local, possibly implemented as just another field in the domain-local state.

> Domain.at_startup (fun () -> Domain.at_exit f)

Seems like the right way to solve the specific use case of the API use in `Format` module with domain-local `at_exit` function.

**Comment by Octachron** [comment]
A possible way to hide internal functions is to not move them to a private section of the documentation using a
`(**/**)`  marker : https://ocaml.org/manual/ocamldoc.html#ss:ocamldoc-stop .

**Comment by gasche** [comment]
I agree that a domain-local `Domain.at_exit` makes more sense. But that creates an asymmetry with `Domain.at_startup` which is similiarly named and has a global semantics (on each future spawn). Sorry for the nitpicking, but do people have a naming suggestion that clarifies the difference? (`Domain.at_each_spawn`, in tandem with `Domain.at_first_spawn`? But this does not suggest that the callback runs on the child.).

**Comment by kayceesrk** [comment]
`at_each_spawn` as a name is sensible. It avoids the unintended and incorrect symmetry between `at_startup` and `at_exit`. Also, I'm not opposed to hiding these functions from the public interface.

**Comment by dbuenzli** [comment]
Following the discussion only from far but note that as far as the design of domain life-cycle is concerned it would also worth having a look at  https://github.com/ocaml/ocaml/issues/11074. The whole thing seems rather unconvincing at the moment (especially the semantics of `Domain.join` w.r.t. to exceptions which looks unusable to me). Besides similar ambiguities (does it affect the current thread of execution/all future ones) exist for uncaught exceptions handlers which would be worth clarifying and design in hand with what will end up being proposed here.

**Comment by kayceesrk** [comment]
@dbuenzli responded in #11074. 

@gasche how do we make progress on this?

**Comment by gasche** [comment]
I think a good first step would be to mark the controversial functions "experimental" / "for internal use" for now.¹ We should, at the same time, refine the API according to feedback.

¹: one problem with just "making them internal" is  that those functions were introduced to make global-state libraries domain-safe in a performance-preserving way, and they would rather naturally be useful to other users (than Format) for the same purpose.

**Comment by kayceesrk** [comment]
Ok. It looks like working on a PR based on the feedback is a good thing to start with. I'll have a go at this.

**Comment by kayceesrk** [comment]
I believe the original problem mentioned in this issue is solved by #11213. Hence, closing this issue.
